File: monitor/central/tree/tree_view.py
from    PyQt5.QtCore   import (
    Qt,
    QSize,
    QSortFilterProxyModel,
    QSettings
)

from    PyQt5.QtWidgets    import (
    QTreeView,
    QMenu,
    QAction,
    QAbstractItemView
)

from    functools import partial
from    monitor.commands.wizards           import UserActionsWizard
from    monitor.logwin.main                import openLoggerFor
from    monitor.central.tree.tree_model    import ProbeModel
from    monitor.central.tree.menus.probe   import ProbeMenu
from    monitor.central.tree.menus.target  import TargetMenu
import  monitor.central.tree.tree_delegate as pdelegate
import  monitor.api                        as monapi
import  sysmapi

class ProbesTreeview(QTreeView):
    def __init__(self, parent):
        super(ProbesTreeview, self).__init__(parent)
        sysmapi.nConnectWillClose(self._saveSettings)
        ProbesTreeview.singleton = self
        self._viewDialogs = dict()
        
        # QTreeView conf
        self.setItemsExpandable(False)
        self.setRootIsDecorated(False)
        self.setWordWrap(True)
        self.setVerticalScrollMode(QAbstractItemView.ScrollPerPixel)
        self.setSelectionMode(QAbstractItemView.SingleSelection)
        self.setHorizontalScrollBarPolicy(Qt.ScrollBarAlwaysOff)
        self.setEditTriggers(QAbstractItemView.NoEditTriggers)
        self.setContextMenuPolicy(Qt.CustomContextMenu)
        self.customContextMenuRequested.connect(self._showMenu)
        self.setObjectName('backTree')
        self.setIconSize(QSize(16, 16)) 
        self.setAnimated(False)
        self.setAlternatingRowColors(True)
        self.setStyleSheet('''QFrame#backTree { 
            background-image: url(./graphics/hover_info_files.png);
            background-repeat: no-repeat;
            background-color: %s;
            background-attachment: fixed; 
            background-position: bottom right}''' % sysmapi.nGetRgb('Base'))
        self.setSortingEnabled(True)
        self.doubleClicked.connect(self._doubleClicked)


        # model conf
        self.model   = ProbeModel(self)
        self.proxy   = QSortFilterProxyModel(self)
        self.proxy.setFilterCaseSensitivity(Qt.CaseInsensitive)
        self.proxy.setDynamicSortFilter(True)
        self.proxy.setSourceModel(self.model)
        self.proxy.setFilterRole(Qt.UserRole)
        self.setModel(self.proxy)


        # delegates
        #self.setItemDelegateForColumn(1, pdelegate.LoggerItemDelegate(self))
        #self.setItemDelegateForColumn(1, pdelegate.ProgressItemDelegate(self))
        #self.setItemDelegateForColumn(2, pdelegate.StatusItemDelegate(self))
        #self.setItemDelegateForColumn(3, pdelegate.TriggerItemDelegate(self))
        #self.setItemDelegateForColumn(4, pdelegate.StateItemDelegate(self))
        #self.setItemDelegateForColumn(5, pdelegate.HostItemDelegate(self))
        #self.setItemDelegateForColumn(7, pdelegate.TimelineItemDelegate(self))

        # setings
        settings = QSettings()
        state   = settings.value('monitor/probe_treeview_header')
        if state != None:
            header = self.header()
            # TOPYQT ERROR BEGIN
            # TypeError: QHeaderView.restoreState(QByteArray): argument 1 has unexpected type 'QVariant'
            #header.restoreState(state)
            # TOPYQT ERROR END

        # various init
        self.targetMenu = TargetMenu(self)
        self.probeMenu  = ProbeMenu(self)
        #self._initMenus()

    def _saveSettings(self):
        header = self.header()
        settings = QSettings()
        settings.setValue('monitor/probe_treeview_header', header.saveState())

    def _doubleClicked(self, index):
        probeItem = index.sibling(index.row(), 0)
        if probeItem.data(Qt.UserRole + 1) == "Probe":
            probe = probeItem.data(Qt.UserRole + 3)
            display = probeItem.data(Qt.DisplayRole)
            openLoggerFor(probe, display)

    def filterThis(self, text):
        self.proxy.setFilterFixedString(text)
        self.expandAll()
                
    def getSelectedElements(self):
        proxyIndexes = self.selectedIndexes()
        selectionList = list()
        for i in range(len(proxyIndexes)):
            modelIndex = self.proxy.mapToSource(proxyIndexes[i])
            modelItem  = self.model.itemFromIndex(modelIndex)
            selectionList.append(modelItem.name)
        return selectionList

    def _showMenu(self, point):
        index = self.proxy.mapToSource(self.indexAt(point))
        item  = self.model.itemFromIndex(index)
        if      item.__class__.__name__ == 'ProbeItem':
            self.probeMenu.showMenuFor(item.name, point)
        elif    item.__class__.__name__ == 'TargetItem':
            self.targetMenu.showMenuFor(item.name, point)
        elif    item.__class__.__name__ == 'NoneType': pass

File: monitor_tree/treeview.cpp
#include "treeview.h"

TreeView::TreeView(QWidget* parent) : QTreeView(parent)
{
    this->setItemsExpandable(true);
    this->setRootIsDecorated(false);
    this->setWordWrap(true);
    this->setVerticalScrollMode(QAbstractItemView::ScrollPerPixel);
    this->setSelectionMode(QAbstractItemView::SingleSelection);
    this->setEditTriggers(QAbstractItemView::NoEditTriggers);
    this->setContextMenuPolicy(Qt::CustomContextMenu);
    QObject::connect(
                this, SIGNAL(customContextMenuRequested(QPoint)),
                this, SLOT(openContextMenu(QPoint)));
    this->setObjectName("MonitorTreeView");
    this->setIconSize(QSize(22,22));
    this->setAnimated(true);
    this->setAlternatingRowColors(true);
    this->setAllColumnsShowFocus(false);
    this->setSortingEnabled(true);
    this->setExpandsOnDoubleClick(false);

    this->original_model = new TreeModel(this);
    this->filter_model   = new QSortFilterProxyModel(this);
    this->filter_model->setSourceModel(this->original_model);
    this->filter_model->setDynamicSortFilter(true);
    this->filter_model->setFilterCaseSensitivity(Qt::CaseSensitive);
    this->filter_model->setFilterRole(Sysmo::ROLE_FILTER_STRING);
    this->setModel(this->filter_model);

    this->timer = new QTimer(this);
    this->timer->setInterval(1000);
    this->timer->setSingleShot(false);
    this->timer->start();
    DelegateProbeProgress* progress = new DelegateProbeProgress(this);

    QObject::connect(
                this->original_model, SIGNAL(expandIndex(QModelIndex)),
                this,                 SLOT(expandIndex(QModelIndex)));
    /*
     * connect delegate first
     */
    QObject::connect(
                this->timer, SIGNAL(timeout()),
                progress,    SLOT(ticTimeout()));

    QObject::connect(
                this->timer,      SIGNAL(timeout()),
                this->viewport(), SLOT(update()));

    this->setItemDelegateForColumn(3, progress);

    QFile file(":/tree/tree.qss");
    file.open(QIODevice::ReadOnly | QIODevice::Text);
    this->setStyleSheet(file.readAll());
    file.close();

    /*
     * Acceptable defaults. Overriden by restore state.
     */
    this->setColumnWidth(0, 250);
    this->setColumnWidth(1, 65);
    this->setColumnWidth(2, 65);
    this->setColumnWidth(3, 100);

    // TODO restore state
}


TreeView::~TreeView()
{
    /*
     * save state
     */
}


// SLOTS
void TreeView::expandIndex(QModelIndex index)
{
    this->expand(this->filter_model->mapFromSource(index));
}

void TreeView::stopTimer() { this->timer->stop(); }

void TreeView::openContextMenu(const QPoint point) {
    QModelIndex    index = this->filter_model->mapToSource(this->indexAt(point));
    QStandardItem* item  = this->original_model->itemFromIndex(index);
    if (item->type() == Sysmo::TYPE_PROBE) {
        QString element_name = item->data(Sysmo::ROLE_ELEMENT_NAME).toString();
        QJsonObject val = Monitor::probe_map->value(element_name);
        qDebug() << "probe name: " << element_name << val;
    } else if (item->type() == Sysmo::TYPE_TARGET) {
        QString element_name = item->data(Sysmo::ROLE_ELEMENT_NAME).toString();
        QJsonObject val = Monitor::target_map->value(element_name);
        qDebug() << "target name: " << element_name << val;
    }
}
